### ISSUE TITLE
Local SQL on different server

### ISSUE THREAD

**Issue by lukas-kaspar** [microsoft/navcontainerhelper#1412]
**Issue Description**

Absolute Docker beginner here, sorry in advance!

I am trying to connect a BC container to a local SQL server which is installed on `Computer B`. 

The docker container is being created on `Computer A`

![Untitled Diagram (3)](https://user-images.githubusercontent.com/7881214/97525065-1cc31f00-19fa-11eb-8672-89df32157e63.png)

As described in [Freddy's blog post](https://freddysblog.com/2019/11/04/using-sql-server-on-the-host/) I am trying to use `host.containerherlper.internal` but I presume that is pointing to `Computer A`. 

I have tried to:

- Replace `host.containerherlper.internal` with the servername `Computer B`
- Replace `host.containerherlper.internal` with the IP of `Computer B`

I feel like I'm missing something basic here.

The db login `docker` has systemadmin role, I can login via SSMS and it's fine. The SQL Server is using a default instance. I've tried using an empty string `''` in `$databaseInstance` with no luck.

Any help would be appreciated.

**Scripts used to create container and cause the issue**
```
$containerName = 'DevNav2017'
$credential = Get-Credential -Message 'Using Windows authentication. Please enter your Windows credentials for the host computer.'
$auth = 'Windows'
$artifactUrl = Get-NavArtifactUrl -nav '2017' -cu '41' -country 'au'
$databaseServer = 'host.containerhelper.internal'
$databaseInstance = 'MSSQLSERVER'
$databaseName = 'DevNav2017'
$databaseUsername = 'docker'
$databasePassword = 'password'
$databaseSecurePassword = ConvertTo-SecureString -String $databasePassword -AsPlainText -Force
$databaseCredential = New-Object pscredential $databaseUsername, $databaseSecurePassword
$licenseFile = 'license.flf'
New-BcContainer `
    -accept_eula `
    -containerName $containerName `
    -credential $credential `
    -auth $auth `
    -artifactUrl $artifactUrl `
    -databaseServer $databaseServer -databaseInstance $databaseInstance -databaseName $databaseName `
    -databaseCredential $databaseCredential `
    -licenseFile $licenseFile `
    -includeCSIDE -doNotExportObjectsToText `
    -updateHosts
```

**Full output of scripts**
```
PS C:\Users\lukask\Desktop> .\createContainer.ps1
BcContainerHelper is version 1.0.8
BcContainerHelper is running as administrator
Host is Microsoft Windows Server 2016 Datacenter - ltsc2016
Docker Client Version is 19.03.12
Docker Server Version is 19.03.12
Fetching all docker images
Using image mcr.microsoft.com/dynamicsnav:10.0.14393.3866-generic
Creating Container DevNav2017
Version: 10.0.30219.0-au
Style: onprem
Multitenant: No
Platform: 10.0.30219.0
Generic Tag: 0.1.0.24
Container OS Version: 10.0.14393.3866 (ltsc2016)
Host OS Version: 10.0.14393.3866 (ltsc2016)
Using process isolation
Using locale en-AU
Disabling the standard eventlog dump to container log every 2 seconds (use -dumpEventLog to enable)
Using license file \\fltmelwfs04\y_drive\Admin\License Files\BC14\BC14 20200909.flf
Files in C:\ProgramData\BcContainerHelper\Extensions\DevNav2017\my:
- AdditionalOutput.ps1
- AdditionalSetup.ps1
- license.flf
- MainLoop.ps1
- SetupVariables.ps1
- updatehosts.ps1
Creating container DevNav2017 from image mcr.microsoft.com/dynamicsnav:10.0.14393.3866-generic
cf15df9f314a5273497a245acf7f93cd608253a5889301b0829a0aa6ce5304da
Waiting for container DevNav2017 to be ready
Using artifactUrl https://bcartifacts.azureedge.net/onprem/10.0.30219.0/au
Using installer from C:\Run\100
Copying Platform Artifacts
Copying Application Artifacts
Copying Installers
Copying TestToolKit
Copying UpgradeToolKit
Copying Extensions
Installing NAV
Starting Local SQL Server
Installing Url Rewrite
Installing SQL Clr Types
Installing Report Viewer
Installing OpenXML
Starting Internet Information Server
Copying Service Tier Files
Copying Web Client Files
Copying Windows Client Files
Copying PowerShell Scripts
Copying ConfigurationPackages
Copying Test Assemblies
Copying TestToolKit
Copying UpgradeToolKit
Copying Extensions
Copying ClientUserSettings
Copying ReportBuilder
Skipping restore of Cronus database
Installing C:\NAVDVD\Installers\AU\OlAddin\OutlookAddIn.Local.AU.msi
Installing C:\NAVDVD\Installers\AU\RTC\RoleTailoredClient.Local.AU.msi
Installing C:\NAVDVD\Installers\AU\Server\Server.Local.AU.msi
Installing C:\NAVDVD\Installers\AU\WebClient\WebClient.Local.AU.msi
Skipping C:\NAVDVD\Installers\AU\WebHelp\WebHelp.Local.AU.msi
Modifying NAV Service Tier Config File for Docker
Creating NAV Service Tier
Installing SIP crypto provider: 'C:\Windows\System32\NavSip.dll'
Installation took 123 seconds
Installation complete
Initializing...
Setting host.containerhelper.internal to 172.30.208.1 in container hosts file
Starting Container
Hostname is DevNav2017
PublicDnsName is DevNav2017
Using Windows Authentication
Import Encryption Key
Cannot establish a connection to the SQL Server/Database.

This could be due to one of the following reasons:

* SQL Server is not started.
* The database does not exist.
* Proper permissions have not been given to the NAV Server Account.

Try again later or contact your system administrator.
at <ScriptBlock>, C:\Run\SetupDatabase.ps1: line 148
at <ScriptBlock>, C:\Run\navstart.ps1: line 113
at <ScriptBlock>, C:\Run\start.ps1: line 349
at <ScriptBlock>, <No file>: line 1Error
Using artifactUrl https://bcartifacts.azureedge.net/onprem/10.0.30219.0/au
Using installer from C:\Run\100
Copying Platform Artifacts
Copying Application Artifacts
Copying Installers
Copying TestToolKit
Copying UpgradeToolKit
Copying Extensions
Installing NAV
Starting Local SQL Server
Installing Url Rewrite
Installing SQL Clr Types
Installing Report Viewer
Installing OpenXML
Starting Internet Information Server
Copying Service Tier Files
Copying Web Client Files
Copying Windows Client Files
Copying PowerShell Scripts
Copying ConfigurationPackages
Copying Test Assemblies
Copying TestToolKit
Copying UpgradeToolKit
Copying Extensions
Copying ClientUserSettings
Copying ReportBuilder
Skipping restore of Cronus database
Installing C:\NAVDVD\Installers\AU\OlAddin\OutlookAddIn.Local.AU.msi
Installing C:\NAVDVD\Installers\AU\RTC\RoleTailoredClient.Local.AU.msi
Installing C:\NAVDVD\Installers\AU\Server\Server.Local.AU.msi
Installing C:\NAVDVD\Installers\AU\WebClient\WebClient.Local.AU.msi
Skipping C:\NAVDVD\Installers\AU\WebHelp\WebHelp.Local.AU.msi
Modifying NAV Service Tier Config File for Docker
Creating NAV Service Tier
Installing SIP crypto provider: 'C:\Windows\System32\NavSip.dll'
Installation took 123 seconds
Installation complete
Initializing...
Setting host.containerhelper.internal to 172.30.208.1 in container hosts file
Starting Container
Hostname is DevNav2017
PublicDnsName is DevNav2017
Using Windows Authentication
Import Encryption Key
Cannot establish a connection to the SQL Server/Database.

This could be due to one of the following reasons:

* SQL Server is not started.
* The database does not exist.
* Proper permissions have not been given to the NAV Server Account.

Try again later or contact your system administrator.
at <ScriptBlock>, C:\Run\SetupDatabase.ps1: line 148
at <ScriptBlock>, C:\Run\navstart.ps1: line 113
at <ScriptBlock>, C:\Run\start.ps1: line 349
at <ScriptBlock>, <No file>: line 1
Initialization of container DevNav2017 failed
At C:\Program Files\WindowsPowerShell\Modules\BcContainerHelper\1.0.8\ContainerHandling\Wait-NavContainerReady.ps1:43
char:17
+ ...             throw "Initialization of container $containerName failed" ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Initialization ...vNav2017 failed:String) [], RuntimeException
    + FullyQualifiedErrorId : Initialization of container DevNav2017 failed
```

**Comment by freddydk** [comment]
host.containerhelper.internal will and should always point to the host,
You can create (on the host) an entry in hosts called: db.containerhelper.internal with the IP address of the database server and use db.containerhelper.internal as your database server.
All entries matching *.containerhelper.internal will be copied from the host to the container at startup.

**Comment by lukas-kaspar** [comment]
That makes a lot of sense - thank you for the prompt response and explanation.

It doesn't seem to want to copy the newly added host entry to the container. When I run `New-BcContainer` it only writes the Docker Virtual Network Gateway:

`Setting host.containerhelper.internal to 172.30.208.1 in container hosts file`

This is my local hosts file:

`PS C:\Users\lukask> cat C:\Windows\System32\drivers\etc\hosts`
```
# Copyright (c) 1993-2009 Microsoft Corp.
#
# This is a sample HOSTS file used by Microsoft TCP/IP for Windows.
#
# This file contains the mappings of IP addresses to host names. Each
# entry should be kept on an individual line. The IP address should
# be placed in the first column followed by the corresponding host name.
# The IP address and the host name should be separated by at least one
# space.
#
# Additionally, comments (such as these) may be inserted on individual
# lines or following the machine name denoted by a '#' symbol.
#
# For example:
#
#      102.54.94.97     rhino.acme.com          # source server
#       38.25.63.10     x.acme.com              # x client host
# localhost name resolution is handled within DNS itself.
#       127.0.0.1       localhost
#       ::1             localhost
192.168.50.98       db.containerhelper.internal
```

This is what's in the container:

`[DEVNAV2017]: PS C:\Run> cat C:\Windows\System32\drivers\etc\hosts`

```
# Copyright (c) 1993-2009 Microsoft Corp.
#
# This is a sample HOSTS file used by Microsoft TCP/IP for Windows.
#
# This file contains the mappings of IP addresses to host names. Each
# entry should be kept on an individual line. The IP address should
# be placed in the first column followed by the corresponding host name.
# The IP address and the host name should be separated by at least one
# space.
#
# Additionally, comments (such as these) may be inserted on individual
# lines or following the machine name denoted by a '#' symbol.
#
# For example:
#
#      102.54.94.97     rhino.acme.com          # source server
#       38.25.63.10     x.acme.com              # x client host
# localhost name resolution is handled within DNS itself.
#       127.0.0.1       localhost
#       ::1             localhost
172.30.208.1 host.containerhelper.internal
```

This is the content of `SetupVariables.ps1`

```
# Invoke default behavior
          . (Join-Path $runPath $MyInvocation.MyCommand.Name)
        

. (Join-Path $PSScriptRoot "updatehosts.ps1") -hostsFile "c:\driversetc\hosts"
```

It looks like the path for the hosts file is incorrect?

**Comment by freddydk** [comment]
driversetc inside the container is mapped (--volume) to your ..\drivers\etc on the host.
https://github.com/microsoft/navcontainerhelper/blob/a014d488bc9022d30638788987c4a300dca44f1a/ContainerHandling/New-NavContainer.ps1#L1437

This is what you should see;
![image](https://user-images.githubusercontent.com/10775043/97531387-66891680-19b4-11eb-86d9-e5d1cbedf357.png)

By adding these lines to my hosts
10.10.10.10 db.containerhelper.internal
10.10.10.10 db.docker.internal

**Comment by freddydk** [comment]
After creating the container you can open a prompt inside (using enter-bccontainer) and do cd driversetc to see that you are in the right location

**Comment by lukas-kaspar** [comment]
I found the culprit. It was the spaces in the host hosts file. I copied and modified one of the Microsoft examples which includes all these spaces:

![image](https://user-images.githubusercontent.com/7881214/97639325-7119dd80-1a92-11eb-9189-b9ecd5602b74.png)

As soon as I remove all the spaces, the entry was picked up and copied to the container.

Thank you very much for your help and all your work on this project!
